compiler/next/test/resolution/testScopeResolve.cpp

```diff
@@ -0,0 +1,263 @@
+  std::string contents = "module M {\n"
+                         "  var y;\n"
+                         "  x;\n"
+                         "}\n";
+  setFileText(context, path, contents);
+
+  const ModuleVec& vec = parse(context, path);
+  assert(vec.size() == 1);
+  const Module* m = vec[0]->toModule();
+  assert(m);
+  assert(m->numStmts() == 2);
+  const Variable* x = m->stmt(0)->toVariable();
+  assert(x);
+  const Identifier* xIdent = m->stmt(1)->toIdentifier();
+  assert(xIdent);
+
+
+  const Scope* scopeForIdent = scopeForId(context, xIdent->id());
+  assert(scopeForIdent);
+  
+  auto pair = findInnermostDecl(context, scopeForIdent, xIdent->name());
+  assert(pair.first == ID());
+  assert(pair.second == 0);
+}
+
+// testing duplicate matching variable declarations
+static void test3() {
+  printf("test3\n");
+  Context ctx;
+  Context* context = &ctx;
+
+  auto path = UniqueString::build(context, "test3.chpl");
+  std::string contents = "module M {\n"
+                         "  var x;\n"
+                         "  var x;\n"
+                         "  x;\n"
+                         "}\n";
+  setFileText(context, path, contents);
+
+  const ModuleVec& vec = parse(context, path);
+  assert(vec.size() == 1);
+  const Module* m = vec[0]->toModule();
+  assert(m);
+  assert(m->numStmts() == 3);
+  const Variable* x1 = m->stmt(0)->toVariable();
+  assert(x1);
+  const Variable* x2 = m->stmt(1)->toVariable();
+  assert(x2);
+  const Identifier* xIdent = m->stmt(2)->toIdentifier();
+  assert(xIdent);
+
+  const Scope* scopeForIdent = scopeForId(context, xIdent->id());
+  assert(scopeForIdent);
+  
+  auto pair = findInnermostDecl(context, scopeForIdent, xIdent->name());
+  assert(pair.first == x1->id());
+  assert(pair.second == 2);
+}
+
+// testing a simple use statement
+static void test4() {
+  printf("test4\n");
+  Context ctx;
+  Context* context = &ctx;
+
+  auto path = UniqueString::build(context, "test4.chpl");
+  std::string contents = "module M {\n"
+                         "  var x;\n"
+                         "}\n"
+                         "module N {\n"
+                         "  use M;\n"
+                         "  x;\n"
+                         "}\n";
+  setFileText(context, path, contents);
+
+  const ModuleVec& vec = parse(context, path);
+  assert(vec.size() == 2);
+  const Module* m = vec[0]->toModule();
+  assert(m);
+  assert(m->numStmts() == 1);
+  const Module* n = vec[1]->toModule();
+  assert(n);
+  assert(n->numStmts() == 2);
+
+  const Variable* x = m->stmt(0)->toVariable();
+  assert(x);
+
+  const Identifier* xIdent = n->stmt(1)->toIdentifier();
+  assert(xIdent);
+
+  const Scope* scopeForIdent = scopeForId(context, xIdent->id());
+  assert(scopeForIdent);
+  
+  auto pair = findInnermostDecl(context, scopeForIdent, xIdent->name());
+  assert(pair.first == x->id());
+  assert(pair.second == 1);
+}
+
+// testing a simple recursive use statement
+static void test5() {
```

Comment:
Chapel supports mutually recursive modules to a point - it's fine for functions because those can get called at any time (and potentially even class inheritance?  I'd have to experiment to be certain), but for things like variables that are initialized when the module is first used or imported, we can't have those rely on something in a module that uses it.  E.g. the following would definitely fail
```chapel
module A {
  use B;
  var x = y;
}
module B {
  use A;
  var y = x;
}
```